Add : 진입화면 및 회원가입 UI 구현 완료

minju1009 opened the pull request:
### PR Type
- [ ] 버그 수정
- [ ] 신규 기능
- [x] 기타 (기타인 경우 내용 입력 : UI구현 ) 

### 해당 PR의 목적
- 진입화면 및 회원가입 UI 구현

### 해당 PR 사유
- 위와 같습니다.

### 리뷰어가 중점적으로 봤으면 하는 부분
1. flex를 올바르게 사용하고 있는지 확인부탁드립니다.
   - 전체 Screen을 하나 만들어 `flex:1`을 주고 자식 컴포넌트들은 비율(`flex:5, flex:1 등`)로 분할해 주었습니다.
2. 아래 사진의 회색 bar가 status bar를 의미하는 것이라 생각되어, 전체 Screen 컴포넌트에 해당 bar의 높이만큼 padding을 주었습니다. 아래 그림의 회색 bar가 status bar을 의미하는게 맞는지 확인차 여쭙습니다.

[zeplin]
<img width="400" alt="image" src="https://user-images.githubusercontent.com/90900882/165436152-78390e56-fd04-4ab6-ba84-8995856ea96f.png">

[구현화면]
<img width="400" alt="image" src="https://user-images.githubusercontent.com/90900882/165436616-b5fcbd6e-176e-492b-860b-389e23b8e6da.png">

minju1009 committed:
Add : 진입화면 및 회원가입 UI 구현 완료

minju1009 committed:
Add : 진입화면 및 회원가입 UI 구현 완료
- android 버튼 크기, 이미지 import 및 경로설정, react navigation headers 사용으로 변경
- 공통 버튼 한 개로 통일하여 활용

jasonkang14 commented on src/screens/Entry/index.tsx:
```diff
@@ -1,15 +1,70 @@
+  Signup: undefined;
+};
+
+type Props = NativeStackScreenProps<RootStackParamList, 'Entry'>;
```
굳굳 좋습니다 👍

minju1009 replied:
감사합니다!

jasonkang14 commented on src/screens/Entry/index.tsx:
```diff
@@ -1,15 +1,70 @@
+    <ScreenContainer>
+      <LogoContainer>
+        <Logo
+          source={require('/Users/kimminju/Desktop/puzzleAI/Puzzleai/src/assets/images/logo_o.png')}
```
typescript를 사용하시면 이미지 파일도 npm package와 유사하게 import 해서 사용할 수 있습니다. 
자료를 찾아보시면 타입을 설정하는 방법과 import하는 방법을 찾으실 수 있을거에요. 
지금 경로는 다른 개발자들의 컴퓨터에는 존재하지 않아 에러가 발생하게 됩니다.

minju1009 replied:
1. 프로젝트 최상단 root에 'global.d.ts' 파일 생성 후 
2. 모듈을 선언해주고
3. 상대 경로로 image를 import 해서 사용해 주었습니다.

jasonkang14 replied:
아직 push를 안하신 것 같습니다! 일단 방향은 맞는데요, 
typescript에서 d.ts파일은 [Module.d.ts](https://www.typescriptlang.org/docs/handbook/declaration-files/templates/module-d-ts.html) 입니다. 
`global`은 일반적으로 전역 변수를 나타내기 때문에, typescript에서 타입을 정하기에는 적합하지는 않습니다
지금 하고자하는 부분에서 `module`은 `image`가 됩니다.

minju1009 replied:
아직 다른 문제들이 해결되지 않아서 push를 안했었는데요! 본의아니게 번거롭게 해드렸네요 ^^; 
알려주셔서 감사합니다. image.d.ts로 수정하였습니다!

jasonkang14 commented on src/screens/Signup/index.tsx:
```diff
@@ -1,14 +1,150 @@
+      <Header>
+        <SignupText>회원가입</SignupText>
+        <GoBackBtn
+          source={require('/Users/kimminju/Desktop/puzzleAI/Puzzleai/src/assets/images/arrow_left.png')}
```
여기도 이미지 경로를 require대신 import할 수 있도록 수정해주세요

jasonkang14 replied:
`GoBackBtn` 보다는 [react navigation headers](https://reactnavigation.org/docs/headers/)를 설정하는 편이 더 일반적입니다.

minju1009 replied:
아이폰 기준으로 보다보니, headers와 다르게 생겨 별도로 만들어줘야 하는 줄 알았습니다 ^^; 안드로이드랑 ios 두 개 다 열어놓고 만들어야 했는데 놓쳤네요!! 

두 개 다 열어 작업 중이고, 말씀해주신 `react navigation headers` 활용하여 수정 후, android의 경우 default가 left이길래 center로 수정해주었습니다 :)

jasonkang14 commented on src/screens/Signup/index.tsx:
```diff
@@ -1,14 +1,150 @@
+          <Title>이메일</Title>
+          <InputBox
+            placeholder="이메일을 입력해주세요"
+            keyboardType="email-address"
```
굳굳 좋습니다 👍

minju1009 replied:
👍

jasonkang14 commented on src/screens/Entry/index.tsx:
```diff
@@ -1,15 +1,70 @@
+        />
+      </LogoContainer>
+      <BtnContainer>
+        <LoginBtn />
```
로그인 버튼은 컴포넌트로 따로 생성하고 회원가입 버튼은 인라인인 이유가 있나요?

minju1009 replied:
로그인 / 회원가입 / 가입완료 세 가지의 버튼이 다 기본적으로 같은데 색이랑 텍스트만 바꿔주고 싶었는데,
방법을 몰라 가장 많이 쓰이는 로그인 버튼만 컴포넌트화 해주었었습니다 😭

타입스크립트 + 스타일드 컴포넌트에서 props로 스타일 내려주는 방법 찾아서 적용하였고,
`LoginSignupBtn`이라는 하나의 공통 컴포넌트를 만들어 props를 통해 모두 적용해 주었습니다!

jasonkang14 reviewed (commented):
flexbox 활용은 좋습니다!
하지만 ios와 안드로이드를 빌드해보시면 버튼 크기가 다른걸 확인하실 수 있으실거에요. 
슬랙에서 질문하신 iOS, android 두 플랫폼 중에 하나만 적용되는 prop이 필요합니다.

minju1009 committed:
Fix : Conflict 해결

minju1009 committed:
Fix : globa.d.ts -> image.d.ts로 수정

minju1009 commented:
피드백 주신 사항에 대해 comment 단 것 외에, 가장 위에서 말씀해 주신 버튼 크기도 수정완료하였습니다.
다먄, 몇 시간동안 찾아봤으나 slack으로 질문드렸던 onPress에 대한 부분이 해결되지 않고 있습니다. 
적용해 보았던 방법은 아래와 같습니다.

1. `onPress={onPress}`로 연결해 주고 따로 함수 선언한 후, onPress를 props로 전달받는 `LoginSignupBtn`에 아래와 같이 타입 할당해주기 
https://stackoverflow.com/a/59901968 
2. `Navigator.tsx`에서 정의해 준 `RootStackParamList`를 export한 후 useNavigator를 사용하기
https://stackoverflow.com/questions/68779417/navigation-navigatehome-showing-some-error-in-typescript
3. type에 any 넣어서 시도해보기
4. typescript 버전 업그레이드 하기

그 외에도 여러 방법을 찾아 시도해보았는데 해결할 수 없었습니다.
React Navigation과 Typescript가 미숙하여, 제가 찾은 해결방법을 이상하게 적용한 건지 몰라 우선 PR 올립니다.
도움 부탁드립니다!

jasonkang14 commented on android/build.gradle:
```diff
@@ -26,7 +26,7 @@ buildscript {
     }
     dependencies {
-        classpath("com.android.tools.build:gradle:7.0.4")
+        classpath('com.android.tools.build:gradle:7.1.3')
```
버전 업그레이드 이유는 무엇인가요?

minju1009 replied:
안드로이드 앱 빌드 과정에서 오류가 나서 찾아보니 버전 업그레이드가 필요하다고 해서 업그레이드 해주었습니다.

jasonkang14 commented on image.d.ts:
```diff
@@ -0,0 +1,4 @@
+declare module '*.png' {
```
추후에 custom type을 선언할 경우를 대비해서 `types` 디렉토리를 생성해서 모아두는 편이 좋을 것 같습니다.

minju1009 replied:
네, 디렉토리 생성 후 해당 파일 넣어주었습니다.

jasonkang14 commented on src/components/LoginSignupBtn.tsx:
```diff
@@ -0,0 +1,55 @@
+  bgColor: string;
+  textColor: string;
+  borderColor?: string;
+  onPress?: any;
```
1. any의 사용은 최대한 지양해주세요. 
2. 버튼에 `onPress` prop이 없는 경우는 어떤 상황인가요?

minju1009 replied:
1. 네, 알겠습니다.
2. 해당 버튼을 회원가입 페이지에서도 사용 중인데, 아직 회원가입 페이지의 버튼 연결을 해주지 않아 오류가 발생하여 우선은 ?로 넣어두었습니다. 버튼 연결 후 수정하도록 하겠습니다 :)
3. 회원가입 페이지에 있는 버튼의 경우 추후 조건에 따라 색을 변경시켜 줄 것이기 때문에해당 페이지에서 관리하는게 더 나을 것 같아 방향 수정하였습니다!

jasonkang14 commented on src/components/LoginSignupBtn.tsx:
```diff
@@ -0,0 +1,55 @@
+}
+
+interface ButtonProps {
+  bgColor: string;
```
color에 `"afliehufwawE"`와 같은 이상한 스트링이 넘어오는 것을 방지하기 위해서, 아래와같이 타입을 별도로 지정할 수도 있습니다. 
```typescript
interface ButtonProps {
  bgColor: "#ffffff" | "#777777";
  borderColor: "#036ad1" | "#g087ga"
}
```

jasonkang14 commented on src/components/LoginSignupBtn.tsx:
```diff
@@ -0,0 +1,55 @@
+
+const Text = styled.Text<TextProps>`
+  ${Platform.select({
+    ios: css`
```
`css` 기능을 이미 알고 계시는군요. 
그렇다면 `borderColor`, `bgColor`, `textColor`를 prop으로 넘기는 것 보다, 
로그인 버튼, 회원가입 버튼 등의 스타일을 `styled-components`의 `css`로 선언해두고, 
`login`, `signup` 을 prop으로 넘겨주면 
부모 component에서 넘어오는 prop이 줄어들어 코드가 더 깔끔해 질 수 있습니다.

minju1009 replied:
`css` 기능을 명확히 알지 못했는데, 덕분에 찾아보고 적용할 수 있었습니다. 확실히 `prop`이 줄어드니까 훨씬 보기 깔끔한 것 같습니다!

jasonkang14 commented on src/screens/Entry/index.tsx:
```diff
@@ -1,15 +1,65 @@
-// type Props = {};
-// const Entry = (props: Props) => {
-const Entry = () => {
+type RootStackParamList = {
```
다양한 화면에서 사용되는 type일 것 같습니다. `types`라는 디렉토리를 생성해서 `image.d.ts`와 navigation 관련 type들을 모아두는 편이 더 좋을 것 같습니다.

minju1009 replied:
`types`디렉토리 생성 후, `image.d.ts`와 `navigation`파일 생성하여 넣어두었습니다! navigation의 경우 아래 두 가지를 정리해 두었습니다.

1. RootStackParamList 
2. 추후 useNavigate 사용을 대비해 default로 타입지정 하기 위한 선언

리액트 네비게이션 [공식문서](https://reactnavigation.org/docs/typescript/#specifying-default-types-for-usenavigation-link-ref-etc)를 보고 따라해 보았는데, 혹시 잘못되었다면 말씀 부탁드립니다!

jasonkang14 commented on src/screens/Entry/index.tsx:
```diff
@@ -1,15 +1,65 @@
+type EntryProps = NativeStackScreenProps<RootStackParamList, 'Entry'>;
+
+const Entry = ({navigation}: EntryProps) => {
+  const onPress = (path: any) => {
```
1. 버튼의 이벤트 핸들링 하는 함수의 이름을 조금 더 명확하게 작성해주세요
2. `any`의 사용은 최대한 지양해주세요
3.  지금과 같은 상황에서는`path`라는 parameter로 숫자, 문자열, 함수, 배열, 객체 등 다양한 값들이 올 수 있습니다. `RootStackParamList`에서 설정한 값들만 parameter로 넘어올 수 있도록 타입을 지정해주세요

minju1009 replied:
1. 네, 알겠습니다!
2. 확인했습니다. 전반적으로 제가 styled-Components를 사용하면서 헷갈려 `onPress`를 잘못된 위치에 사용하면서 type도 이상하게 넣었었네요 ^^; `any`의 경우 타입스크립트를 사용하는 이점이 없어지는 것이나 마찬가지기에 사용을 지양해야 하는 것으로 이해하였습니다. 
3. 해당 페이지에서는 `Login`과 `Signup`두 가지만 넘어와야 하기에 `(path: 'Login' | 'Signup')`와 같이 수정해주었습니다.

jasonkang14 commented on src/components/LoginSignupBtn.tsx:
```diff
@@ -0,0 +1,55 @@
+  textColor: string;
+}
+
+const LoginSignupBtn = ({
```
`text` 대신 `children` prop을 사용할 수 있도록 변경해주세요.

minju1009 replied:
`children` prop 사용으로 변경해주었습니다.

jasonkang14 commented on src/screens/Entry/index.tsx:
```diff
@@ -1,15 +1,65 @@
+          textColor="white"
+          onPress={onPress('Login')}
+        />
+        <LoginSigunupBtn
```
`LoginSignupBtn`에 했던 comment를 모두 반영하면 아래와 같이 간단해질 수 있습니다.
```typescript
<LoginSignupBtn id="signup" onPress={() => handleSignupBtnPress()}>회원가입</LoginSignupBtn>
```

minju1009 replied:
확인 후 변경하였습니다 :) 감사합니다~!

jasonkang14 commented on src/screens/Entry/index.tsx:
```diff
@@ -1,15 +1,65 @@
+import styled from 'styled-components/native';
+import type {NativeStackScreenProps} from '@react-navigation/native-stack';
+import LoginSigunupBtn from 'components/LoginSignupBtn';
+import logo from './../../assets/images/logo.png';
```
image도 절대경로를 사용해주세요

minju1009 replied:
네, 절대경로로 수정하였습니다.

jasonkang14 commented:
> 

질문에 답변을 안했네요 ㅠㅠ
1번이 가장 적절해보입니다. 
저는 개인적으로 `screen`에서 `useNavigation()`의 사용을 지양하고자 하는 편인데요, 
왜냐면 Stack에 포함된 모든 screen은 navigation을 prop으로 받기 때문입니다. 

일반적으로 웹에서 보면 event가 발생하지만, 
그 이벤트를 사용할 때만 explicit하게 Event를 적어줍니다. 
react native도 유사하다고 보시면 됩니다. 
따라서 함수를 따로 선언하는 경우에는 이벤트를 넘겨줄 수 있지만 넘겨주지 않는? 
다고 보시면 될 것 같습니다. 
답변이 너무 늦어져서 죄송합니다.

minju1009 committed:
"Fix : 피드백 사항 반영"
- 1. onPress 함수 적용 위치 수정
- 2. styled-component의 css기능 사용하여 버튼 prop 줄여줌
- 3. types 디렉토리 생성 후 image.d.ts 및 navigation type 파일 정리
- 4. 이미지 절대경로로 수정

minju1009 commented:
답변 감사드립니다! 
`screen`은 navigation을 prop으로 받기 때문에 `NativeStackScreenProps`을 활용하여 navigation을 사용한다는 점 공식문서를 통해 추가 이해하였습니다! 

styled component를 사용하다보니 함수 적용위치를 이상하게 적용하고 문제의 원인을 다른데에서 찾고 있었네요. 그렇지만 그 덕에 typescript 포함 여러가지를 공부해 볼 수 있었다고 생각합니다 :) 
함께 찾아봐주시고 도움 주셔서 감사합니다!

jasonkang14 commented on src/components/LoginSignupBtn.tsx:
```diff
@@ -0,0 +1,38 @@
+
+interface LoginSignupBtnProps {
+  id: string;
+  children: JSX.Element;
```
완전 꼼꼼하게 다 적어주셨네요 고생하셨습니다! 👍

jasonkang14 commented on src/components/LoginSignupBtn.tsx:
```diff
@@ -0,0 +1,38 @@
+const LoginSignupBtn = ({navigate, children, id}: LoginSignupBtnProps) => {
+  return (
+    <Button id={id} onPress={navigate}>
+      {children}
```
💯

jasonkang14 commented on src/components/LoginSignupBtn.tsx:
```diff
@@ -0,0 +1,38 @@
+  width: 100%;
+  margin-bottom: 15px;
+  border-radius: 8px;
+  ${({id}) =>
```
스타일 코드가 너무 길어진다고 보실 수도 있지만, 
추상화라는 측면에서는 이런식으로 처리하는 편이 더 좋은 것 같습니다.
내부에 작성하셔도 좋지만 따로 변수를 선언할 수도 있습니다. 
```typescript
const LoginButton = css`
  background-color: ${({theme}) => theme.primary};
`

const SignupButton = css`
  background-color: ${({theme}) => theme.white};
  border: 1px solid ${({theme}) => theme.primary};
`

const Button = styled.TouchableOpacity<ButtonProps>`
...
${(props) => props.id === 'login' ? LoginButton : SignupButton}
`
```

minju1009 replied:
오..!!! 위와 같이 작성하는게 훨씬 가독성이 좋네요!! 감사합니다! 👍

jasonkang14 reviewed (approved):
고생하셨습니다!
질문이 하나 있는데
`/types/navigation.tsx`에서 `ReactNavigation`에 관한 타입도 지정을 해줘야 하던가요?

minju1009 commented:
여기에도 답변 달아놓습니다 :)

말씀하신대로 타입을 따로 선언하지 않아도 작동은 되는데 [useNavigation 타입지정](https://reactnavigation.org/docs/typescript/#annotating-usenavigation)에 관한 공식문서를 보니 완벽한 type-safe한 방법이 아니라서, useNavigation을 자주 사용할거면 [defaut type으로 지정](https://reactnavigation.org/docs/typescript/#specifying-default-types-for-usenavigation-link-ref-etc)하는 방법에 따라 지정해주면 일일히 type지정을 해주지 않아도 될 뿐더러, type-safety도 지켜준다고 합니다!

실제로 코드에 적용해보니, default로 적용해주니 문제 없이 잘 되고, default로 적용해주지 않으면 별도의 type지정이 필요합니다!
